File: CS/MessageBoxHelper/MyMessageBox/MyXtraMessageBoxForm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.XtraEditors;

namespace MessageBoxHelper.Internal {
    public class MyXtraMessageBoxForm : DevExpress.XtraEditors.XtraMessageBoxForm {
        private Timer CountdownTimer;
        private bool countDownCaption;
        private bool autoClose;
        private int timeout;
        private DialogResult timeoutResult;
        private string baseCaption;
        private bool captionWasCoped;
        private bool selectNext;
        private bool disablebttns;
        private bool disableCancel;
        private bool center;
        public bool ShowMessageNextTime;

        private CheckEdit cbShowNextTime;
        public MyXtraMessageBoxForm()
            : base() {
            baseCaption = "";
            CountdownTimer = new Timer();
            CountdownTimer.Tag = this;
            CountdownTimer.Interval = 1000;
            CountdownTimer.Tick += MyTimer_Tick;
            captionWasCoped = false;
        }

        protected override void OnShown(EventArgs e) {            
            if (countDownCaption)
                this.Text = "Remain " + timeout + " sec";
            if (selectNext) {
                ShowNextTime();
            }
            IButtonControl btnCancel = null;
            if ((this.CancelButton != null) && (this.CancelButton.DialogResult == DialogResult.Cancel))
                btnCancel = this.CancelButton;
            if (disablebttns)
                DisableAllExceptCancel(btnCancel as SimpleButton);
            if ((disableCancel) && (btnCancel != null))
                (btnCancel as SimpleButton).Enabled = false;
            if (center)
                CenterOnParent();
            if (this.timeout > 0) {
                this.CountdownTimer.Enabled = true;
                this.CountdownTimer.Start();
            }
            base.OnShown(e);
        }

        void DisableAllExceptCancel(SimpleButton btnCancel) {
            foreach (Control cont in this.Controls)
                if ((cont is BaseButton) && (cont != (btnCancel as SimpleButton)))
                    cont.Enabled = false;
        }

        void MyTimer_Tick(object sender, EventArgs e) {
            if (!captionWasCoped) {
                baseCaption = this.Message.Caption;
                captionWasCoped = true;
            }
            Timer messageTimer = (sender as Timer);
            timeout--;
            if ((countDownCaption) && (timeout > 0)) {
                this.Text = "Remain " + timeout + " sec";
            } else if (countDownCaption)
                this.Text = baseCaption;
            if (timeout <= 0) {
                this.Message.Caption = baseCaption;
                messageTimer.Stop();
                messageTimer.Enabled = false;
                if (autoClose) {
                    this.DialogResult = timeoutResult;
                    base.Close();
                }
            }
        }

        public DialogResult ShowForm(MyXtraMessageArgs messageArgs) {
            selectNext = messageArgs.ShowNextTime;
            disablebttns = messageArgs.DisableButtons;
            disableCancel = messageArgs.DisableCancel;
            autoClose = messageArgs.AutoClose;
            countDownCaption = messageArgs.ShowCountdown;
            timeout = messageArgs.Timeout;
            center = messageArgs.Center;
            timeoutResult = messageArgs.AutoCloseResult;

            return base.ShowMessageBoxDialog(messageArgs);
        }

        protected override void OnClosing(System.ComponentModel.CancelEventArgs e) {
            if (cbShowNextTime != null)
                this.ShowMessageNextTime = cbShowNextTime.Checked;
            this.CountdownTimer.Stop();
            this.CountdownTimer.Dispose();
            base.OnClosing(e);
        }

        void ShowNextTime() {
            cbShowNextTime = new CheckEdit();
            cbShowNextTime.Checked = true;
            cbShowNextTime.Text = "Show this dialog again?";
            cbShowNextTime.Properties.AutoWidth = true;
            cbShowNextTime.Properties.AutoHeight = true;
            this.Height += cbShowNextTime.Height;
            cbShowNextTime.Location = new Point(1, this.ClientSize.Height - cbShowNextTime.Height - 1);
            this.Controls.Add(cbShowNextTime);
        }

        void CenterOnParent() {
            int y = this.Owner.Height - this.Height;
            int x = this.Owner.Width - this.Width;
            this.Location = new Point(x / 2 + this.Owner.Location.X, y / 2 + this.Owner.Location.Y);
        }
    }
}

File: VB/MessageBoxHelper/MyMessageBox/MyXtraMessageBoxForm.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Drawing
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.XtraEditors

Namespace MessageBoxHelper.Internal
    Public Class MyXtraMessageBoxForm
        Inherits DevExpress.XtraEditors.XtraMessageBoxForm

        Private CountdownTimer As Timer
        Private countDownCaption As Boolean
        Private autoClose As Boolean
        Private timeout As Integer
        Private timeoutResult As DialogResult
        Private baseCaption As String
        Private captionWasCoped As Boolean
        Private selectNext As Boolean
        Private disablebttns As Boolean
        Private disableCancel As Boolean
        Private center As Boolean
        Public ShowMessageNextTime As Boolean

        Private cbShowNextTime As CheckEdit
        Public Sub New()
            MyBase.New()
            baseCaption = ""
            CountdownTimer = New Timer()
            CountdownTimer.Tag = Me
            CountdownTimer.Interval = 1000
            AddHandler CountdownTimer.Tick, AddressOf MyTimer_Tick
            captionWasCoped = False
        End Sub

        Protected Overrides Sub OnShown(ByVal e As EventArgs)
            If countDownCaption Then
                Me.Text = "Remain " & timeout & " sec"
            End If
            If selectNext Then
                ShowNextTime()
            End If
            Dim btnCancel As IButtonControl = Nothing
            If (Me.CancelButton IsNot Nothing) AndAlso (Me.CancelButton.DialogResult = DialogResult.Cancel) Then
                btnCancel = Me.CancelButton
            End If
            If disablebttns Then
                DisableAllExceptCancel(TryCast(btnCancel, SimpleButton))
            End If
            If (disableCancel) AndAlso (btnCancel IsNot Nothing) Then
                TryCast(btnCancel, SimpleButton).Enabled = False
            End If
            If center Then
                CenterOnParent()
            End If
            If Me.timeout > 0 Then
                Me.CountdownTimer.Enabled = True
                Me.CountdownTimer.Start()
            End If
            MyBase.OnShown(e)
        End Sub

        Private Sub DisableAllExceptCancel(ByVal btnCancel As SimpleButton)
            For Each cont As Control In Me.Controls
                If (TypeOf cont Is BaseButton) AndAlso (cont IsNot (TryCast(btnCancel, SimpleButton))) Then
                    cont.Enabled = False
                End If
            Next cont
        End Sub

        Private Sub MyTimer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            If Not captionWasCoped Then
                baseCaption = Me.Message.Caption
                captionWasCoped = True
            End If
            Dim messageTimer As Timer = (TryCast(sender, Timer))
            timeout -= 1
            If (countDownCaption) AndAlso (timeout > 0) Then
                Me.Text = "Remain " & timeout & " sec"
            ElseIf countDownCaption Then
                Me.Text = baseCaption
            End If
            If timeout <= 0 Then
                Me.Message.Caption = baseCaption
                messageTimer.Stop()
                messageTimer.Enabled = False
                If autoClose Then
                    Me.DialogResult = timeoutResult
                    MyBase.Close()
                End If
            End If
        End Sub

        Public Function ShowForm(ByVal messageArgs As MyXtraMessageArgs) As DialogResult
            selectNext = messageArgs.ShowNextTime
            disablebttns = messageArgs.DisableButtons
            disableCancel = messageArgs.DisableCancel
            autoClose = messageArgs.AutoClose
            countDownCaption = messageArgs.ShowCountdown
            timeout = messageArgs.Timeout
            center = messageArgs.Center
            timeoutResult = messageArgs.AutoCloseResult

            Return MyBase.ShowMessageBoxDialog(messageArgs)
        End Function

        Protected Overrides Sub OnClosing(ByVal e As System.ComponentModel.CancelEventArgs)
            If cbShowNextTime IsNot Nothing Then
                Me.ShowMessageNextTime = cbShowNextTime.Checked
            End If
            Me.CountdownTimer.Stop()
            Me.CountdownTimer.Dispose()
            MyBase.OnClosing(e)
        End Sub

        Private Sub ShowNextTime()
            cbShowNextTime = New CheckEdit()
            cbShowNextTime.Checked = True
            cbShowNextTime.Text = "Show this dialog again?"
            cbShowNextTime.Properties.AutoWidth = True
            cbShowNextTime.Properties.AutoHeight = True
            Me.Height += cbShowNextTime.Height
            cbShowNextTime.Location = New Point(1, Me.ClientSize.Height - cbShowNextTime.Height - 1)
            Me.Controls.Add(cbShowNextTime)
        End Sub

        Private Sub CenterOnParent()
            Dim y As Integer = Me.Owner.Height - Me.Height
            Dim x As Integer = Me.Owner.Width - Me.Width
            Me.Location = New Point(x \ 2 + Me.Owner.Location.X, y \ 2 + Me.Owner.Location.Y)
        End Sub
    End Class
End Namespace
